File: packages/sdk/src/webhook.ts
/**
 * Hookwing Webhook verifier
 *
 * Verifies HMAC-SHA256 signatures on incoming Hookwing webhook deliveries.
 * Works in Node.js (v18+), browsers, and edge runtimes (Cloudflare Workers, etc.)
 *
 * @example
 * ```ts
 * import { Webhook } from '@hookwing/sdk';
 *
 * const wh = new Webhook('whsec_your_signing_secret');
 *
 * // In your webhook handler:
 * const payload = await request.text();
 * const signature = request.headers.get('X-Hookwing-Signature');
 * const timestamp = request.headers.get('X-Hookwing-Timestamp');
 *
 * const event = await wh.verify(payload, { signature, timestamp });
 * // event is typed as WebhookEvent
 * ```
 */

export const SIGNATURE_HEADER = 'x-hookwing-signature';
export const TIMESTAMP_HEADER = 'x-hookwing-timestamp';
export const EVENT_TYPE_HEADER = 'x-event-type';

/** Default tolerance window for timestamp verification (5 minutes) */
const DEFAULT_TOLERANCE_MS = 5 * 60 * 1000;

export interface WebhookEvent {
  id: string;
  type: string;
  timestamp: number;
  payload: unknown;
}

export interface VerifyHeaders {
  signature: string | null;
  timestamp: string | null;
}

export class WebhookVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'WebhookVerificationError';
  }
}

export class Webhook {
  private readonly secret: string;
  private readonly toleranceMs: number;

  constructor(secret: string, options?: { toleranceMs?: number }) {
    if (!secret) {
      throw new Error('Webhook secret is required');
    }
    this.secret = secret.startsWith('whsec_') ? secret.slice(6) : secret;
    this.toleranceMs = options?.toleranceMs ?? DEFAULT_TOLERANCE_MS;
  }

  /**
   * Verify a webhook payload and return the parsed event.
   * Throws WebhookVerificationError if the signature is invalid or the timestamp is stale.
   */
  async verify(payload: string, headers: VerifyHeaders): Promise<WebhookEvent> {
    const { signature, timestamp } = headers;

    if (!signature) {
      throw new WebhookVerificationError('Missing X-Hookwing-Signature header');
    }
    if (!timestamp) {
      throw new WebhookVerificationError('Missing X-Hookwing-Timestamp header');
    }

    // Validate timestamp format and staleness
    const ts = Number(timestamp);
    if (Number.isNaN(ts)) {
      throw new WebhookVerificationError('Invalid X-Hookwing-Timestamp header');
    }

    const now = Date.now();
    const age = Math.abs(now - ts);
    if (age > this.toleranceMs) {
      throw new WebhookVerificationError(
        `Webhook timestamp too old (${Math.round(age / 1000)}s ago, tolerance: ${Math.round(this.toleranceMs / 1000)}s)`,
      );
    }

    // Verify HMAC signature
    const isValid = await this.verifySignature(payload, signature);
    if (!isValid) {
      throw new WebhookVerificationError('Webhook signature verification failed');
    }

    // Parse payload
    let parsed: unknown;
    try {
      parsed = JSON.parse(payload);
    } catch {
      throw new WebhookVerificationError('Webhook payload is not valid JSON');
    }

    const event = parsed as Record<string, unknown>;

    return {
      id: typeof event.id === 'string' ? event.id : '',
      type: typeof event.type === 'string' ? event.type : 'unknown',
      timestamp: ts,
      payload: event,
    };
  }

  /**
   * Verify signature only (without timestamp check).
   * Use `verify()` in production — this is for testing.
   */
  async verifySignature(payload: string, signatureHeader: string): Promise<boolean> {
    if (!signatureHeader.startsWith('sha256=')) {
      return false;
    }

    const expectedSig = signatureHeader.slice(7);
    const actualSig = await this.computeHmac(payload);

    // Constant-time comparison
    if (expectedSig.length !== actualSig.length) {
      return false;
    }

    let result = 0;
    for (let i = 0; i < expectedSig.length; i++) {
      result |= (expectedSig.codePointAt(i) ?? 0) ^ (actualSig.codePointAt(i) ?? 0);
    }
    return result === 0;
  }

  private async computeHmac(payload: string): Promise<string> {
    const encoder = new TextEncoder();
    const keyData = encoder.encode(this.secret);
    const messageData = encoder.encode(payload);

    const key = await crypto.subtle.importKey(
      'raw',
      keyData,
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign'],
    );
    const signature = await crypto.subtle.sign('HMAC', key, messageData);
    const hashArray = Array.from(new Uint8Array(signature));
    return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
  }
}

File: packages/go-sdk/webhook.go
package hookwing

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"
)

const (
	// SignatureHeader is the header name for the webhook signature.
	SignatureHeader = "X-Hookwing-Signature"
	// TimestampHeader is the header name for the webhook timestamp.
	TimestampHeader = "X-Hookwing-Timestamp"
	// EventTypeHeader is the header name for the event type.
	EventTypeHeader = "X-Event-Type"

	// DefaultToleranceMs is the default tolerance for timestamp verification (5 minutes).
	DefaultToleranceMs int64 = 5 * 60 * 1000
)

// WebhookVerificationError represents an error during webhook verification.
type WebhookVerificationError struct {
	Message string
}

func (e *WebhookVerificationError) Error() string {
	return e.Message
}

// WebhookOption is a functional option for configuring a Webhook.
type WebhookOption func(*Webhook)

// WithToleranceMs sets the tolerance for timestamp verification in milliseconds.
func WithToleranceMs(toleranceMs int64) WebhookOption {
	return func(w *Webhook) {
		w.toleranceMs = toleranceMs
	}
}

// Webhook verifies HMAC-SHA256 signatures on incoming webhook deliveries.
type Webhook struct {
	secret      string
	toleranceMs int64
}

// NewWebhook creates a new Webhook verifier.
func NewWebhook(secret string, opts ...WebhookOption) *Webhook {
	// Strip 'whsec_' prefix if present
	secret = strings.TrimPrefix(secret, "whsec_")

	w := &Webhook{
		secret:      secret,
		toleranceMs: DefaultToleranceMs,
	}

	for _, opt := range opts {
		opt(w)
	}

	return w
}

// Verify verifies a webhook payload and returns the parsed event.
func (w *Webhook) Verify(payload []byte, headers http.Header) (*WebhookEvent, error) {
	signature := headers.Get(SignatureHeader)
	timestamp := headers.Get(TimestampHeader)

	if signature == "" {
		return nil, &WebhookVerificationError{Message: "Missing X-Hookwing-Signature header"}
	}
	if timestamp == "" {
		return nil, &WebhookVerificationError{Message: "Missing X-Hookwing-Timestamp header"}
	}

	// Validate timestamp format
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return nil, &WebhookVerificationError{Message: "Invalid X-Hookwing-Timestamp header"}
	}

	// Check timestamp staleness
	now := time.Now().UnixMilli()
	age := now - ts
	if age < 0 {
		age = -age
	}
	if age > w.toleranceMs {
		return nil, &WebhookVerificationError{
			Message: fmt.Sprintf("Webhook timestamp too old (%ds ago, tolerance: %ds)",
				age/1000, w.toleranceMs/1000),
		}
	}

	// Verify HMAC signature
	if !w.VerifySignature(payload, signature) {
		return nil, &WebhookVerificationError{Message: "Webhook signature verification failed"}
	}

	// Parse payload
	var parsed map[string]any
	if err := json.Unmarshal(payload, &parsed); err != nil {
		return nil, &WebhookVerificationError{Message: "Webhook payload is not valid JSON"}
	}

	// Determine event type from header or payload
	eventType := headers.Get(EventTypeHeader)
	if eventType == "" {
		if t, ok := parsed["type"].(string); ok {
			eventType = t
		} else {
			eventType = "unknown"
		}
	}

	// Get event ID from payload
	eventID := ""
	if id, ok := parsed["id"].(string); ok {
		eventID = id
	}

	return &WebhookEvent{
		ID:        eventID,
		Type:      eventType,
		Timestamp: ts,
		Payload:   parsed,
	}, nil
}

// VerifySignature verifies the signature only (without timestamp check).
func (w *Webhook) VerifySignature(payload []byte, signatureHeader string) bool {
	if !strings.HasPrefix(signatureHeader, "sha256=") {
		return false
	}

	expectedSig := strings.TrimPrefix(signatureHeader, "sha256=")
	actualSig := w.computeHMAC(payload)

	// Constant-time comparison
	expectedBytes, err := hex.DecodeString(expectedSig)
	if err != nil {
		return false
	}
	actualBytes, err := hex.DecodeString(actualSig)
	if err != nil {
		return false
	}

	return hmac.Equal(expectedBytes, actualBytes)
}

// computeHMAC computes the HMAC-SHA256 of the payload.
func (w *Webhook) computeHMAC(payload []byte) string {
	h := hmac.New(sha256.New, []byte(w.secret))
	h.Write(payload)
	return hex.EncodeToString(h.Sum(nil))
}
